### PR TITLE
[EICNET-2081] my notifications management

### PR DESCRIPTION
**Pull request by Askerx9**
### What I did
**_Be sure to use freshly created groups & events (the follow flag has been changed for global & group events, for old groups you have to unfollow/follow again)_**
- [x] Go to /my-settings
- [x] Be sure your profile is filled (especially with topics & regions)
- [x] You should see your topics & region within the "Interest notifications" tab
- [x] Create an event within a group (should be followed automatically)
- [x] Create a global event (should be followed automatically)
- [x] Click on toggles within the interest, groups, events & comments tabs
- [x] Refresh the page & your changes should be applied
- [x] Try unsubscribing from some of the entities you follow, this should unfollow that and the row shouldn't appear anymore

### Toggle checkbox

- http://localhost:8080/themes/custom/eic_community/storybook-static/?path=/story/components-toogle-switch--base
> Now we have a component that we can reuse in some Forms

### My settings layout
- StoryBook: http://localhost:8080/themes/custom/eic_community/storybook-static/?path=/story/bundles-member-my-settings--email-notifications
- Layout: https://www.figma.com/file/j583oU3vO3YfrU3SnZp4vQ/EIC---Analysis?node-id=335%3A11398

### My setting/notifications in twig
- StoryBook: http://localhost:8080/themes/custom/eic_community/storybook-static/?path=/story/compositions-notifications-management-line--base
- Layout: https://www.figma.com/file/j583oU3vO3YfrU3SnZp4vQ/EIC---Analysis?node-id=374%3A14153

**Remarks**
> this is a static view (no collapse) because everything is done in React part

### My setting/notifications in React
- Layout: https://www.figma.com/file/j583oU3vO3YfrU3SnZp4vQ/EIC---Analysis?node-id=374%3A14153

> to test, put this code in html.html.twig
`   
 {{ attach_library('eic_community/react-notif-management') }}
    <div id="notif-management-overview"></div>
`

### CODE REVIEW

File: lib/themes/eic_community/react/components/Block/Toggle/entrypoint.js [hunk 1]
```diff
@@ -0,0 +1,13 @@
+import React from "react";
+import ReactDOM from "react-dom";
+
+import Toggle from "./index";
```
Comment (Dromadoros) [inline lib/themes/eic_community/react/components/Block/Toggle/entrypoint.js:4]: I think the empty line should be below the import Toggle

File: lib/themes/eic_community/react/components/Block/NotifManagement/index.js [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+        }.bind(this))
+        .catch(function (error) {
+          self.setState({
+            loading: false
+          });
+        });
+    }
+  }
+  changePage(page) {
+    console.log('navigate to page', page);
```
Comment (Dromadoros) [inline lib/themes/eic_community/react/components/Block/NotifManagement/index.js:108]: Do we need this on environments?

File: lib/themes/eic_community/react/components/Block/NotifManagement/index.js [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+        'notification_status': 'Status',
+        'notification_unsubscribe': 'Unsubscribe',
+        'toggle_on': 'On',
+        'toggle_off': 'Off',
+        'no_results_title': 'No results',
+        'no_results_body': 'No results body goes here'
+      },
+      initiated: false,
+      total: this.props.demo ? 80 : 0,
+      resultsPerPage: 10,
```
Comment (Dromadoros) [inline lib/themes/eic_community/react/components/Block/NotifManagement/index.js:42]: isn't it better to receive this value from the back-end? Like Overviews we have

File: lib/themes/eic_community/react/components/Block/NotifManagement/components/RowNotif.js [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+  setShowChildren(state) {
+    this.setState({
+      showChildren: state
+    });
+  }
+  render() {
+    return (
+      <>
+        {this.props.hidden !== true &&
+          <Row translations={this.props.translations} parentIndex={this.props.rowIndex} row={this.props.row}
```
Comment (Dromadoros) [inline lib/themes/eic_community/react/components/Block/NotifManagement/components/RowNotif.js:28]: Can we maybe indent this better, props line by line? just for readability

File: lib/themes/eic_community/react/components/Block/NotifManagement/index.js (hunk 3 above)
Comment (filipvt1) [reply lib/themes/eic_community/react/components/Block/NotifManagement/index.js:42]: yes, when we implement the backend endpoint these values should definitely come from the backend. The fixed total was added for test/demo purposes to show the pagination.

File: lib/themes/eic_community/react/components/Block/NotifManagement/index.js (hunk 2 above)
Comment (filipvt1) [reply lib/themes/eic_community/react/components/Block/NotifManagement/index.js:108]: you mean the console.log? No that should be removed. But we do still need to implement the pagination mechanism here.

File: lib/modules/eic_user/src/NotificationFrequencies.php [hunk 5]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace Drupal\eic_user;
```
Comment (lramarojaona) [inline lib/modules/eic_user/src/NotificationFrequencies.php:3]: Wouldn't it be better to move this into `eic_message_subscriptions` module?

File: lib/modules/eic_user/eic_user.services.yml [hunk 6]
```diff
@@ -10,3 +10,7 @@ services:
       - [ setDocumentProcessor, [ '@?eic_search.solr_document_processor' ] ]
     tags:
       - { name: event_subscriber }
+
+  eic_user.notification_settings_manager:
```
Comment (lramarojaona) [inline lib/modules/eic_user/eic_user.services.yml:14]: Wouldn't it be better to move this into `eic_message_subscriptions` module?

File: lib/modules/eic_user/src/Controller/MySettingsController.php [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+
+  /**
+   * The member activities endpoint.
+   *
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *
+   * @return array
+   */
+  public function settings(Request $request): array {
+    return [];
```
Comment (lramarojaona) [inline lib/modules/eic_user/src/Controller/MySettingsController.php:57]: Is this a required function?

File: lib/modules/eic_user/src/Controller/MySettingsController.php [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+  }
+
+  /**
+   * @param string $notification_type
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *
+   * @return \Symfony\Component\HttpFoundation\JsonResponse
+   */
+  public function setProfileNotificationSettings(string $notification_type, Request $request): JsonResponse {
+    $body = json_decode($request->getContent(), TRUE);
```
Comment (lramarojaona) [inline lib/modules/eic_user/src/Controller/MySettingsController.php:67]: I usually use this https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Component%21Serialization%21Json.php/function/Json%3A%3Adecode/9.3.x

But it does the same I agree

File: lib/modules/eic_user/src/Plugin/Block/NotificationsSettingsBlock.php [hunk 9]
```diff
@@ -0,0 +1,321 @@
...
+            'unsubscribe_modal_confirm' => $this->t('Confirm'),
+            'unsubscribe_modal_cancel' => $this->t('Cancel'),
+            'search_placeholder' => $this->t('Search'),
+            'notification_name' => $this->t('Name'),
+            'notification_status' => $this->t('Status'),
+            'notification_unsubscribe' => $this->t('Unsubscribe'),
+            'toggle_on' => $this->t('On'),
+            'toggle_off' => $this->t('Off'),
+            'no_results_title' => $this->t('No results'),
+            'no_results_body' => $this->t('Could not find any results.'),
```
Comment (lramarojaona) [inline lib/modules/eic_user/src/Plugin/Block/NotificationsSettingsBlock.php:128]: Is there a reason not use https://www.drupal.org/docs/8/api/translation-api/overview#s-context-in-drupalt-and-drupalformatplural ?

React maybe?

File: lib/modules/eic_user/src/Service/NotificationSettingsManager.php [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+   * @param \Drupal\flag\FlaggingInterface|null $flagging
+   * @param $value
+   *
+   * @return bool|null
+   * @throws \Drupal\Core\Entity\EntityStorageException
+   */
+  public function setSettingValue(string $notification_type, $value, ?FlaggingInterface $flagging = NULL): ?bool {
+    $user = User::load($this->currentUser->id());
+    if (!$user instanceof UserInterface) {
+      throw new \InvalidArgumentException('Current user doesn\'t exist');
```
Comment (lramarojaona) [inline lib/modules/eic_user/src/Service/NotificationSettingsManager.php:86]: Drupal coding standards recommend `("Current user doesn't exist")` in this case

**Review by lramarojaona — CHANGES_REQUESTED** [review]
It's working, but I have some comments.

Also we need a followup ticket to handle notification frequency, I'll chekc with Lucas

File: lib/modules/eic_user/src/NotificationFrequencies.php (hunk 5 above)
Comment (korhankale) [reply lib/modules/eic_user/src/NotificationFrequencies.php:3]: Already agreed on the way we'll split this

File: lib/modules/eic_user/src/Controller/MySettingsController.php (hunk 8 above)
Comment (korhankale) [reply lib/modules/eic_user/src/Controller/MySettingsController.php:67]: Indeed it does the same thing (json_decode as array)

File: lib/modules/eic_user/src/Controller/MySettingsController.php (hunk 7 above)
Comment (korhankale) [reply lib/modules/eic_user/src/Controller/MySettingsController.php:57]: Yes in the routing definition it is required that the function exists

**Pull request merged**